lib/debug.ml

```diff
@@ -151,17 +151,23 @@ let set_facility f = Mutex.execute facility_m (fun () -> facility := f)
 let get_facility () = Mutex.execute facility_m (fun () -> !facility)
 
 let output_log brand level priority s =
-  if not(is_disabled brand level) then begin
     let msg = format false brand priority s in
     if !print_debug
     then Printf.printf "%s\n%!" (format true brand priority s);
-
     Syslog.log (get_facility ()) level msg
+
+(* CA-326244: This function takes a lock to check if the debug level is disabled,
+   therefore it shouldn't be called from exception hooks, finalizers, etc. that
+   can get invoked from any place in the code, that might result in a deadlock in this module in
+   case it's invoked while we already hold the loglevel_m mutex *)
+let output_log_if_enabled brand level priority s =
+  if not(is_disabled brand level) then begin
+    output_log brand level priority s
```

Comment:
Is this implementing the same logic twice?

```
let output_log brand level priority s =
  if not(is_disabled brand level) then begin
    let msg = format false brand priority s in
    if !print_debug
    then Printf.printf "%s\n%!" (format true brand priority s);

    Syslog.log (get_facility ()) level msg
  end
```

Reply:
The `if` got moved as part of the PR into the new function.